pkg/backend/display/richDiff.go

```diff
@@ -0,0 +1,175 @@
+	oldParent, newParent resource.PropertyValue) {
+
+	element := path[0]
+
+	old, new := getProperty(element, oldParent), getProperty(element, newParent)
+
+	switch element := element.(type) {
+	case int:
+		if parent.Array == nil {
+			parent.Array = &resource.ArrayDiff{
+				Adds:    make(map[int]resource.PropertyValue),
+				Deletes: make(map[int]resource.PropertyValue),
+				Sames:   make(map[int]resource.PropertyValue),
+				Updates: make(map[int]resource.ValueDiff),
+			}
+		}
+
+		// For leaf diffs, the provider tells us exactly what to record. For other diffs, we will derive the
+		// difference from the old and new property values.
+		if len(path) == 1 {
+			switch kind {
+			case plugin.DiffAdd, plugin.DiffAddReplace:
+				parent.Array.Adds[element] = new
+			case plugin.DiffDelete, plugin.DiffDeleteReplace:
+				parent.Array.Deletes[element] = old
+			case plugin.DiffUpdate, plugin.DiffUpdateReplace:
+				parent.Array.Updates[element] = resource.ValueDiff{Old: old, New: new}
+			default:
+				contract.Failf("unexpected diff kind %v", kind)
+			}
+		} else {
+			switch {
+			case old.IsNull() && !new.IsNull():
+				parent.Array.Adds[element] = new
+			case !old.IsNull() && new.IsNull():
+				parent.Array.Deletes[element] = old
+			default:
+				ed := parent.Array.Updates[element]
+				addDiff(path[1:], kind, &ed, old, new)
+				parent.Array.Updates[element] = ed
+			}
+		}
+	case string:
+		if parent.Object == nil {
+			parent.Object = &resource.ObjectDiff{
+				Adds:    make(resource.PropertyMap),
+				Deletes: make(resource.PropertyMap),
+				Sames:   make(resource.PropertyMap),
+				Updates: make(map[resource.PropertyKey]resource.ValueDiff),
+			}
+		}
+
+		e := resource.PropertyKey(element)
+		if len(path) == 1 {
+			switch kind {
+			case plugin.DiffAdd, plugin.DiffAddReplace:
+				parent.Object.Adds[e] = new
+			case plugin.DiffDelete, plugin.DiffDeleteReplace:
+				parent.Object.Deletes[e] = old
+			case plugin.DiffUpdate, plugin.DiffUpdateReplace:
+				parent.Object.Updates[e] = resource.ValueDiff{Old: old, New: new}
+			default:
+				contract.Failf("unexpected diff kind %v", kind)
+			}
+		} else {
+			switch {
+			case old.IsNull() && !new.IsNull():
+				parent.Object.Adds[e] = new
+			case !old.IsNull() && new.IsNull():
+				parent.Object.Deletes[e] = old
+			default:
+				ed := parent.Object.Updates[e]
+				addDiff(path[1:], kind, &ed, old, new)
+				parent.Object.Updates[e] = ed
+			}
+		}
+	default:
+		contract.Failf("unexpected path element type: %T", element)
+	}
+}
+
+func translateRichDiff(step engine.StepEventMetadata) *resource.ObjectDiff {
+	contract.Assert(step.RichDiff != nil)
+
+	// The rich diff is presented as a list of simple JS property paths and corresponding diffs. We translate this to
+	// an ObjectDiff by progressively inserting ValueDiffs as necessary. Old values are always taken from a step's
+	// Outputs; new values are always taken from its Inputs.
```

Comment:
i'm not certain i understand the comment about "progressively inserting ValueDiffs as necessary"

Reply:
Attempted some clarification.